cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go

```diff
@@ -136,7 +136,14 @@ func newProvider(
 
 // BuildClusterAPI builds CloudProvider implementation for machine api.
 func BuildClusterAPI(opts config.AutoscalingOptions, do cloudprovider.NodeGroupDiscoveryOptions, rl *cloudprovider.ResourceLimiter) cloudprovider.CloudProvider {
-	externalConfig, err := clientcmd.BuildConfigFromFlags("", opts.KubeConfigPath)
+
+	// Backwards compatibility for falling back to opts.KubeConfigPath
+	kubeconfig := opts.CloudConfig
+	if kubeconfig == "" {
+		kubeconfig = opts.KubeConfigPath
+	}
+
```

Comment:
How would this work for the scenario where machines run in the management cluster vs nodes run in the workload cluster? the kubeclient below https://github.com/kubernetes/autoscaler/pull/3203/files#diff-2e4533399e5c94111c2fe5fa4b52dcc9R157 is used to build a node informer so it'd require different kubeconfig for such scenario but all clients here seems to always use `externalConfig`.

Reply:
Ideally longer term, I would like for this to eventually fall back to using an in-cluster config if `--cloud-config` is not specified, but doing so today would introduce a breaking change to users that are relying on the current behavior. I added this mainly to not break existing users until we've given them a chance to migrate to the new usage first.

I'm open to thoughts on how long we need to provide the backwards compatibility and we can add support for the in-cluster config fallback.

Reply:
Yeh the current fallback looks reasonable to me.
My point is that as far as I can see, this PR in its current state won't provide a working autoscaler for the scenario where there's a management cluster (with machines) and a workload cluster (with nodes backing by those machines). Because no matter which flags you set, the code as in the PR is always passing the same kubeconfig to both clients the one which manipulates nodes and the one which manipulates machines https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go#L145-L160

Reply:
@enxebre It doesn't support using the in-cluster config for the client, but it would still support passing in a separate valid kubeconfig configured for the management cluster.

Alternatively, I'm wondering if it would make sense to add an additional boolean flag that can be passed to override the fallback behavior, that would give us the ability to support in-cluster config while not breaking existing users. Thoughts?

Reply:
> but it would still support passing in a separate valid kubeconfig configured for the management cluster.

I'm having a hard time to see how though. Note that the capi provider needs access to both machines(management cluster) and nodes(workload cluster).
The nodes are manipulated by this client https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go#L150

As far as I can see in this PR, The kube client and so the node informer https://github.com/kubernetes/autoscaler/blob/a63c7abbe47773be4eec49b13210a3cd9c4a719f/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go#L372 always uses the same kubeconfig than the client used to manipulates machines i.e `externalConfig`.

Reply:
Ah, you are indeed correct. I'll work on fixing that up. I had overlooked the modifications to the Node on my first pass through.

Reply:
i'm really scratching my head at how this appeared to work for me. i'm guessing just some side-effects of using the docker provider to back capi.

Reply:
In other words I can't see how the capi provider would ever work for the scenario where machines and nodes live in different clusters as far as the kubeclient here receives the same variable than the other clients https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go#L150-L160

Reply:
Yeah, I'll rework things to create two separate clients and use each appropriately

Reply:
Has this been resolved? I think this is what the `/hold` is on right now